### ISSUE TITLE
[overlapped labeling] Multiple output maps.

### ISSUE THREAD

**Issue by moured** [open-mmlab/mmsegmentation#1137]
Hello, 

I have a dataset that has a couple of classes with overlapped labelling. Meaning, a pixel can have multiple segmentations at the same time. I developed a model that can output two maps (I return them as a list [x1 x2]) that is to predict separately every class (as a binary 1/0). 

Now, how can I compute the loss for this model and backpropagate it? 

Note that in forward_train function  mmseg/models/decode_heads/decode_head.py I can see the logits but this line :
losses = self.losses(seg_logits, gt_semantic_seg)
is only working (model is training) when seg_logits is changed to seg_logits[0]

Any ideas ?

Best,
Omar

**Comment by MengzhangLI** [comment]
Hi, sorry for late reply.

Currently, we only support `one pixel` to `one class`. If it is laborious to handle dataset, you can set `ignore_index` on certain classes for normal training.

**Comment by lorinczszabolcs** [comment]
Hi!

This feature would be much appreciated. It could be also handled in a different way, e.g. having multiple heads producing multiple output segmentations, and having multiple targets, one for each head.

Do you have any recommendation where could this be added / modified in the codebase?

**Comment by moured** [comment]
> Hi!
> 
> This feature would be much appreciated. It could be also handled in a different way, e.g. having multiple heads producing multiple output segmentation, and having multiple targets, one for each head.
> 
> Do you have any recommendations on where could this be added/modified in the codebase?


Thanks for your comment. I have done that and currently publishing about it. You can either wait till I push the code soon or follow this flow. You need to change the data loader to read multi-label file structure (image is only 3 channels, you need to choose something else). Then the data augmentation, especially the padding operation needs to be modified and GPU adapted. Finally a customized loss function.

Best,
Omar

**Comment by lorinczszabolcs** [comment]
> > Hi!
> > This feature would be much appreciated. It could be also handled in a different way, e.g. having multiple heads producing multiple output segmentation, and having multiple targets, one for each head.
> > Do you have any recommendations on where could this be added/modified in the codebase?
> 
> Thanks for your comment. I have done that and currently publishing about it. You can either wait till I push the code soon or follow this flow. You need to change the data loader to read multi-label file structure (image is only 3 channels, you need to choose something else). Then the data augmentation, especially the padding operation needs to be modified and GPU adapted. Finally a customized loss function.
> 
> Best, Omar

Hi! Thanks for your reply!

I am also trying to implement the mentioned feature. I figured out the data loader part, I’m still unsure about data augmentation, and will want to implement a multi-head version of the encoder-decoder segmentor and add it to the Segmentors registry by having a list of decode heads instead of having a single one (similarly to how auxiliary heads are inplemented). And would keep the usual losses for each of the heads.

Best,
Szabi

**Comment by mmeendez8** [comment]
Has there been anything new about this?
